src/couch/src/couch_bt_engine_header.erl

```diff
@@ -204,17 +204,16 @@ upgrade_tuple(Old) when is_record(Old, db_header) ->
     Old;
 upgrade_tuple(Old) when is_tuple(Old) ->
     NewSize = record_info(size, db_header),
-    if
-        tuple_size(Old) < NewSize -> ok;
-        true -> erlang:error({invalid_header_size, Old})
-    end,
-    {_, New} = lists:foldl(
-        fun(Val, {Idx, Hdr}) ->
-            {Idx + 1, setelement(Idx, Hdr, Val)}
+    Upgrade = tuple_size(Old) < NewSize,
+    ProhibitDowngrade = config:get_boolean("couchdb", "prohibit_downgrade", true),
+    OldKVs =
+        case {Upgrade, ProhibitDowngrade} of
+            {true, AnyBool} when is_boolean(AnyBool) -> tuple_to_list(Old);
+            {false, true} -> error({invalid_header_size, Old});
+            {false, false} -> lists:sublist(tuple_to_list(Old), NewSize)
```

Comment:
I think this only works if field was added at the end. How about the case when we remove one of the fields in the middle? The downgrade in general is not solvable problem if we don't maintain fields order for a previous version. We can try something like:

```
record_info(1) -> 
     [
        disk_version, update_seq, ...
    ];
...
record_info(7) -> 
     [
        disk_version, update_seq, unused, id_tree_state,
        seq_tree_state, local_tree_state, purge_tree_state,
        purge_seq_tree_state, security_ptr, revs_limit, uuid,
        epochs, compacted_seq, purge_infos_limit, props_ptr
    ];

record_info(?LATEST_DISK_VERSION) ->
     record_info(fields, db_header).
```

If we would maintain such info we should be able to figure out what has been changed.

Reply:
I'll update the commit comment, as it was missing the part that is only meant for cases when the disk version hasn't changed and is meant to work along with the feature were we allowed "easy" upgrades by just appending fields to the header

Reply:
Updated the comment to emphasize it’s for matching disk version numbers only

Reply:
The `NewSize` is greater than the size of the tuple, which means the `lists:sublist` invocation would return the same tuple.

```erlang
Upgrade = tuple_size(Old) < NewSize,
...
{false, false} -> lists:sublist(tuple_to_list(Old), NewSize)
```

Reply:
yeah, and we had handled the `==` as well anyway in the no-op clause of the function:

```
upgrade_tuple(Old) when is_record(Old, db_header) ->
    Old;
```

So only `>` remains and that seems to work like:

```
> lists:sublist(tuple_to_list({rec,a,b}), 2).
[rec,a]
```